lib/Sema/CSSimplify.cpp

```diff
@@ -3939,9 +3940,19 @@ ConstraintSystem::simplifyKeyPathConstraint(Type keyPathTy,
     break;
   }
   
+  // FIXME: Allow the type to be upcast if the type system has a concrete
+  // KeyPath type assigned to the expression already.
+  if (keyPathBGT) {
+    if (keyPathBGT->getDecl() == getASTContext().getKeyPathDecl())
+      kpDecl = getASTContext().getKeyPathDecl();
+    else if (keyPathBGT->getDecl() == getASTContext().getWritableKeyPathDecl()
+             && capability >= Writable)
+      kpDecl = getASTContext().getWritableKeyPathDecl();
+  }
```

Comment:
@rudkx Do you think this is an acceptable incremental improvement for the 4.0 branch in the meantime?